map_find.h

```diff
@@ -0,0 +1,150 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+
+#pragma once
+
+#include <boost/optional.hpp>
+
+namespace navitia {
+namespace utils {
+
+/**
+ * @brief Find key within a map, in a functional way
+ *
+ * The find operation is performed when the object is constructed with a key,
+ * or when find() is called.
+ * Result is accessible via : if_found() or if_not_found()
+ */
+template<class Map>
+class MapFind
+{
+public:
+    typedef typename Map::key_type KeyType;
+    typedef typename Map::mapped_type ValueType;
+    typedef std::reference_wrapper<const ValueType> ConstRefValueType;
+
+    /**
+     * @brief Create a MapFind object on a map, no find is performed until find() is called.
+     *
+     * @param map : the map to perform the search on
+     */
+    MapFind(Map & map): map(map) {}
+
+
+    /**
+     * @brief Create a MapFind object and find the key in the associated map
+     *
+     * @param map : the map to perform the search on
+     * @param key : the key to be searched in the map
+     */
+    MapFind(Map & map, const KeyType & key): map(map) {
+        find(key);
+    }
+
+    /**
+     * @brief Callback fn with the value associated to the searched key
+     * if the find was successful
+     *
+     * @param fn : a callback function that takes a constant mapped type reference as parameter
+     * @returns A MapFind object to query the restult or perfom another find
+     */
+    MapFind& if_found(std::function<void(const ValueType &)> fn) {
+        if( res ) {
+            fn( res->get() );
+        }
+        return *this;
+    }
+
+    /**
+     * @brief Callback fn if nothing was found in the map
+     *
+     * @param fn : a callback function with no parameter
+     * @returns A MapFind object to query the restult or perfom another find
+     */
+    template<class Func>
+    MapFind& if_not_found(Func fn) {
+        if( !res ) {
+            fn();
+        }
+        return *this;
+    }
+
+    /**
+     * @brief Search for key within the map used in the constructor
+     *
+     * @param key : the key to be searched in the map
+     * @returns A MapFind object to query the restult or perfom another find
+     */
+    template<class Key>
+    MapFind& find(const Key & key) {
+        auto it = map.find(key);
+        if(it != map.end()) {
+            auto cref_val = std::cref(it->second);
+            res = boost::make_optional(cref_val);
+        } else {
+            res = boost::none;
+        }
+        return *this;
+    }
+
+private:
+    Map & map;
+    boost::optional<ConstRefValueType> res;
```

Comment:
a boost::optional on a std::reference_wrapper is a bit useless, you can just use a pointer.

Reply:
I agree this means exactly the same thing (and generates the exact same code).
The only thing I'm not fond of if we use a pointer, would be the double deref like:
```cpp
MapFind& if_found(std::function<void(const ValueType &)> fn) {
        if( res ) {
            fn(**res);
        }
[...]
```
But if a pointer makes the code easier to read, I'm happy to change it.